multicast/hear.py

```diff
@@ -0,0 +1,367 @@
+
+
+class HearUDPHandler(socketserver.BaseRequestHandler):
+	"""Subclasses socketserver.BaseRequestHandler for handling echo function.
+
+	Basicly simplifies testing by allowing a trivial echo back (case-insensitive) of string
+	data, after printing the sender's ip out.
+	"""
+
+	def handle(self):
+		data = self.request[0].strip()
+		socket = self.request[1]
+		print(str("{} SAYS: {} to {}").format(
+			self.client_address[0], str(data), "ALL"
+		))
+		if data is not None:
+			myID = str(socket.getsockname()[0])
+			print(
+				str("{me} HEAR: [{you} SAID {what}]").format(
+					me=myID, you=self.client_address, what=str(data)
+				)
+			)
+			print(
+				str("{me} SAYS [ HEAR [ {what} SAID {you} ] from {me} ]").format(
+					me=myID, you=self.client_address, what=str(data)
+				)
+			)
+			send.McastSAY()._sayStep(
+				self.client_address[0], self.client_address[1],
+				str("HEAR [ {what} SAID {you} ] from {me}").format(
+					me=myID, you=self.client_address, what=data.upper()
+				)
+			)
+			if """STOP""" in str(data):
+				raise RuntimeError("SHUTDOWN")
+
+
+class McastHEAR(multicast.mtool):
+	"""Subclasses multicast.mtool to provide the HEAR tooling.
+
+		Testing:
+
+		Testcase 0: First setup test fixtures by importing multicast.
+
+			>>> import multicast
+			>>> multicast.hear is not None
+			True
+			>>> multicast._MCAST_DEFAULT_PORT is not None
+			True
+			>>> multicast._MCAST_DEFAULT_GROUP is not None
+			True
+			>>> multicast._MCAST_DEFAULT_TTL is not None
+			True
+			>>> multicast.hear.McastHEAR is not None
+			True
+			>>>
+
+		Testcase 2: Recv should be detailed with some metadata.
+			A: Test that the __MAGIC__ variables are initialized.
+			B: Test that the __MAGIC__ variables are strings.
+
+			>>> multicast.hear is not None
+			True
+			>>> multicast.hear.McastHEAR is not None
+			True
+			>>> multicast.hear.McastHEAR.__module__ is not None
+			True
+			>>> multicast.hear.McastHEAR.__proc__ is not None
+			True
+			>>> multicast.hear.McastHEAR.__epilogue__ is not None
+			True
+			>>> multicast.hear.McastHEAR.__prologue__ is not None
+			True
+			>>>
+
+
+	"""
+
+	__module__ = """multicast.hear"""
+
+	__name__ = """multicast.hear.McastHEAR"""
+
+	__proc__ = """HEAR"""
+
+	__epilogue__ = """Generally speaking you want to bind to one of the groups you joined in
+		this module/instance, but it is also possible to bind to group which
+		is added by some other programs (like another python program instance of this)
+	"""
+
+	__prologue__ = """Python Multicast Server for multicast input."""
+
+	@classmethod
+	def setupArgs(cls, parser):
+		pass
+
+	def doStep(self, *args, **kwargs):
+		#  _is_std = False if "is_std" not in kwargs.keys() else kwargs["is_std"]
+		HOST = multicast._MCAST_DEFAULT_GROUP if "group" not in kwargs else kwargs["group"]
+		PORT = multicast._MCAST_DEFAULT_PORT if "port" not in kwargs else kwargs["port"]
+		with McastServer((HOST, PORT), HearUDPHandler) as server:
+			server.serve_forever()
```

Comment:
I like the approach to this suggestion, however, first, please refactor your suggestion to use double-quotes `"` instead of single quotes `'` and edit/update both the diff and the committable suggestion.


Second, please also provide a break down with supporting documentation that supports your claim of 

> This change can simplify the code and improve readability without affecting portability or compatibility.

For this break down please consider performance impacts, possible shadowed variables, complexity, and maintainability while continuing to consider readability, portability, and compatibility.


Third, please remind me of what the official python documentation says about the `get()` function and the link where I can find the official documentation, for posterity.

I'm looking forward to this one.

Reply:
Sounds fine to me, let's go with this change (which includes the right quote style)

```suggestion
	def doStep(self, *args, **kwargs):
		#  _is_std = False if "is_std" not in kwargs.keys() else kwargs["is_std"]
		HOST = kwargs.get("group", multicast._MCAST_DEFAULT_GROUP)
+		PORT = kwargs.get("port", multicast._MCAST_DEFAULT_PORT)
		with McastServer((HOST, PORT), HearUDPHandler) as server:
			server.serve_forever()
```